# Delta Lake Time Machine & Optimization Lab

Delta Lake allows you to read, write and query data in data lakes in an efficient manner.

## In this lesson you:
* Compare different versions of a Delta table using Time Machine
* Optimize your Delta Lake to increase speed and reduce number of files

## Audience
* Primary Audience: Data Engineers
* Secondary Audience: Data Analysts and Data Scientists

## Prerequisites
* Web browser: current versions of Google Chrome, Firefox, Safari, Microsoft Edge and
Internet Explorer 11 on Windows 7, 8, or 10 (see <a href="https://docs.databricks.com/user-guide/supported-browsers.html#supported-browsers#" target="_blank">Supported Web Browsers</a>)
* Databricks Runtime 4.2 or greater
## Datasets Used
We will use online retail datasets from `/mnt/training/online_retail`

In [0]:
%run "./Includes/Classroom-Setup"

Initialized classroom variables & functions...

Datasets are already mounted to /mnt/training from wasbs://training@dbtrainwesteurope.blob.core.windows.net/

Created user-specific database

Using the database saifahmed_k_outlook_com_db .

All done!

Because we'll be calculating some aggregates in this notebook, we'll change our partitions after shuffle from the default `200` to `8` (which is a good number for the 8 node cluster we're currently working on).

In [0]:
%python

sqlContext.setConf("spark.sql.shuffle.partitions", "8")

### Check for our previous Delta Lake tables

This lab relies upon some tables created in previous Delta Lake lessons and labs. 

If you get an error from either of the next two SQL queries, running the solution code for the "Delta-Lake-Lab-1" will build all necessary tables.

In [0]:
%sql

SELECT COUNT(*) FROM customer_counts;
SELECT COUNT(*) FROM customer_data_delta;

count(1) 65499

-sandbox
<img alt="Caution" title="Caution" style="vertical-align: text-bottom; position: relative; height:1.3em; top:0.0em" src="https://files.training.databricks.com/static/images/icon-warning.svg"/> **The following cell will take several minutes to execute, and is only necessary to run if you got an error in the previous cell.**

In [0]:
%run "./Includes/Delta-Lab-2-Prep"

For convenience later in this lab, the paths to the files defining our existing Delta tables are provided. You can use these paths to load the data into DataFrames, if desired, though this entire lab can be completed using SQL on the existant tables.

In [0]:
DeltaPath = userhome + "/delta/customer-data/"
CustomerCountsPath = userhome + "/delta/customer_counts/"

**Note: This lab depends upon the complete exectuion of the notebook titled "Open-Source-Delta-Lake" and the "Delta-Lake-Basics" lab. If these tables don't exist, go back and run all cells in these notebook.**

### Time Travel
Because Delta Lake is version controlled, you have the option to query past versions of the data. Let's look at the history of our current Delta table.

In [0]:
%sql
DESCRIBE HISTORY customer_data_delta

version timestamp userId userName operation operationParameters job notebook clusterId readVersion isolationLevel isBlindAppend operationMetrics userMetadata engineInfo 0 2024-10-20T11:57:41.000+0000 7104407279782773 saifahmed.k@outlook.com WRITE Map(mode -> Overwrite, partitionBy -> ["Country"]) null List(2599175015539894) 1019-080424-ac1xqx2j null WriteSerializable false Map(numFiles -> 37, numOutputRows -> 65499, numOutputBytes -> 633565) null Databricks-Runtime/11.3.x-photon-scala2.12

Querying an older version is as easy as adding `VERSION AS OF desired_version`. Let's verify that our table from one version back still exists.

In [0]:
%sql
SELECT COUNT(*)
FROM customer_data_delta
VERSION AS OF 0

count(1) 65499

Using a single file storage system, you now have access to every version of your historical data, ensuring that your data analysts will be able to replicate their reports (and compare aggregate changes over time) and your data scientists will be able to replicate their experiments.

### Check difference between versions

You want to compare how many orders from Sweden were added by your recent UPSERT to your BI table.

Let's start by getting the total sum of our `total_orders` column where our country is Sweden.

In [0]:
# TODO
count = spark.sql(f"SELECT COUNT(*) FROM delta.`{DeltaPath}`").collect()[0][0]

Again, we can look at the history of our Delta table here.

In [0]:
%sql
DESCRIBE HISTORY customer_counts

version timestamp userId userName operation operationParameters job notebook clusterId readVersion isolationLevel isBlindAppend operationMetrics userMetadata engineInfo 1 2024-10-20T12:17:02.000+0000 7104407279782773 saifahmed.k@outlook.com MERGE Map(predicate -> ["((Country#2287 = Country#2212) AND (CustomerID#2286 = CustomerID#2211))"], matchedPredicates -> [{"actionType":"update"}], notMatchedPredicates -> [{"actionType":"insert"}]) null List(930942090424827) 1019-080424-ac1xqx2j 0 WriteSerializable false Map(numTargetRowsCopied -> 4, numTargetRowsDeleted -> 0, numTargetFilesAdded -> 2, executionTimeMs -> 8440, materializeSourceTimeMs -> 893, numTargetRowsInserted -> 4, scanTimeMs -> 5959, numTargetRowsUpdated -> 1205, numOutputRows -> 1213, numTargetChangeFilesAdded -> 0, numSourceRows -> 1209, numTargetFilesRemoved -> 1, rewriteTimeMs -> 1160) null Databricks-Runtime/11.3.x-photon-scala2.12 0 2024-10-20T12:05:38.000+0000 7104407279782773 saifahmed.k@outlook.com WRITE Map(mode -> Overwrite, partitionBy -> []) null List(930942090424827) 1019-080424-ac1xqx2j null WriteSerializable false Map(numFiles -> 1, numOutputRows -> 1209, numOutputBytes -> 8070) null Databricks-Runtime/11.3.x-photon-scala2.12

Our original table will be version `0`. Let's write a SQL query to see how many orders we originally had from Sweden.

In [0]:
%sql
SELECT SUM(total_orders)
FROM customer_counts
VERSION AS OF 0
WHERE Country='Sweden'

sum(total_orders) 41

We can combine these two queries and get our difference, which represents our new entries.

In [0]:
%sql
SELECT SUM(total_orders) - (
  SELECT SUM(total_orders)
  FROM customer_counts
  VERSION AS OF 0
  WHERE Country='Sweden') AS new_entries
FROM customer_counts
WHERE Country='Sweden'

new_entries 41

### OPTIMIZE and ZORDER

Let's apply some of these optimizations to `../delta/customer-data/`.

Our data is partitioned by `Country`.

We want to query the data for `StockCode` equal to `22301`.

We expect this query to be slow because we have to examine ALL OF `../delta/customer-data/` to find the desired `StockCode` and not just in one or two partitions.

First, let's time the above query: you will need to form a DataFrame to pass to `preZorderQuery`.

In [0]:
# TODO
%timeit preZorderQuery = spark.sql(f"SELECT * FROM delta.`{DeltaPath}`").collect()

938 ms ± 76.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Compact the files and re-order by `StockCode`.

In [0]:
%sql
-- TODO
OPTIMIZE customer_counts
ZORDER by (CustomerID)

path metrics dbfs:/user/saifahmed.k@outlook.com/delta/customer_counts List(1, 2, List(8314, 8314, 8314.0, 1, 8314), List(1009, 8314, 4661.5, 2, 9323), 0, List(minCubeSize(107374182400), List(0, 0), List(2, 9323), 0, List(2, 9323), 1, null), 1, 2, 0, false, 0, 0, 1729453315810, 1729453318762, 4, 1, null)

Let's time the above query again: you will need to form a DataFrame to pass to `postZorderQuery`.

In [0]:
# TODO
%timeit postZorderQuery = spark.sql(f"SELECT * FROM delta.`{DeltaPath}`").collect()

844 ms ± 51.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


### OPTIMIZE your BI table

Here we'll optimize our `customer_counts` table so that we can quickly query on our `CustomerID` column.

In [0]:
%sql
OPTIMIZE customer_counts
ZORDER by (CustomerID)

path metrics dbfs:/user/saifahmed.k@outlook.com/delta/customer_counts List(0, 0, List(null, null, 0.0, 0, 0), List(null, null, 0.0, 0, 0), 0, List(minCubeSize(107374182400), List(1, 8314), List(0, 0), 1, List(0, 0), 0, null), 0, 1, 1, false, 0, 0, 1729453394124, 1729453394600, 4, 0, null)

Now we can easily look at which of our customers have made the most orders.

In [0]:
%sql
SELECT CustomerID, SUM(total_orders) AS total
FROM customer_counts
GROUP BY CustomerID
ORDER BY total DESC

CustomerID total null 50562 12748 1390 17841 962 14606 842 15311 836 14911 754 17850 594 13089 522 13081 522 18118 500 15039 494 12681 406 14298 394 14646 368 14505 366 14527 348 13174 338 14667 336 15061 328 17341 322 15574 318 15464 314 18116 310 18041 306 15808 298 12501 298 12433 294 14800 292 14085 288 17511 288 15570 288 15498 288 16033 282 15379 276 17340 274 15044 268 17284 262 14733 248 17338 244 17218 238 15998 236 13488 234 17059 234 14415 234 17406 232 13093 232 15867 230 15640 230 15356 230 15601 230 13069 228 15514 226 17827 224 15005 224 14669 220 12471 218 15880 212 16713 212 17757 210 15023 210 15719 208 17602 206 12451 202 17609 200 13715 200 15727 198 12841 198 13319 196 12484 196 15194 196 16782 194 15555 192 17920 190 16161 190 14083 190 12539 186 16725 184 17894 184 12370 182 13680 182 17611 180 14723 180 17019 180 14796 178 12682 178 15708 178 13694 176 12766 176 14587 176 13458 174 15426 174 16686 174 13798 174 17259 174 17085 172 17231 170 17968 170 17954 168 12472 168 12567 166 15358 166 15034 164 18055 164 13564 164 16931 162 15983 162 17591 160 17961 158 17787 158 14307 158 17365 158 17811 156 17198 154 18069 154 12683 154 16923 154 17722 152 15529 152 14573 152 17690 152 14329 152 14180 152 16904 152 17965 152 17211 150 17372 150 17017 150 16710 150 16065 148 12585 148 17634 148 17450 148 17449 146 17897 146 13767 146 17191 144 18109 144 12637 144 13050 144 12578 142 14729 142 16241 142 14135 140 16016 140 14702 140 13102 140 14472 140 16873 140 15503 140 13764 138 16891 138 14081 138 12383 138 17315 136 13013 134 12577 134 16274 134 16327 132 18043 132 12872 132 16795 132 15889 132 17581 130 17976 130 16919 130 17530 130 15862 128 16814 128 17884 128 15898 128 17890 128 15033 128 14395 126 16244 126 17097 126 13155 126 17126 126 14657 124 15296 122 13124 122 14680 122 16719 120 16395 120 17238 120 17596 120 16779 120 13030 120 13468 120 16255 120 16520 120 12921 118 15159 118 12838 118 14769 118 14099 118 17975 116 14572 116 17908 116 15460 116 17860 116 12867 116 16402 114 15021 114 14544 114 15525 114 18283 114 15002 112 15547 112 15157 112 17243 110 16805 110 18156 110 16270 108 18223 108 15856 108 15012 108 17796 108 14565 108 17303 106 16503 106 12626 106 16791 106 12647 106 15351 106 15673 106 17618 106 17062 106 13982 104 13838 104 17364 104 16670 104 12753 104 13097 104 12540 104 14031 102 14159 102 16550 102 15235 102 13983 100 16916 100 14449 100 15281 100 14985 100 15532 100 14506 100 17091 100 15615 100 17428 100 13418 98 17555 98 16059 98 15581 98 15919 98 17392 98 15024 96 14057 96 12652 96 14907 96 16889 96 13408 96 13451 96 15916 96 17967 96 16746 94 16735 94 17346 94 12415 94 15518 92 16727 92 13267 92 12481 92 12720 92 14156 92 13004 92 14404 92 16442 92 16893 90 17377 90 16718 90 16558 90 13078 90 13370 88 17230 88 14062 88 14684 88 17576 88 14713 88 18168 88 15858 88 13777 86 17551 86 12490 86 17838 86 15485 86 18212 86 12377 86 14215 86 14639 84 12480 84 16722 84 12583 84 17865 84 14092 84 15596 84 17358 84 16029 82 14961 82 13668 82 15723 82 15535 82 15046 80 15271 80 13048 80 18004 80 15502 80 16011 80 16470 80 17950 80 13162 80 14901 80 13246 80 15093 78 15298 78 16422 78 14810 78 17251 78 13593 78 18106 78 14016 78 17675 78 13263 78 16655 78 14978 76 15894 76 14577 76 17912 76 15078 76 13842 76 15716 76 17999 76 17189 76 13634 76 16145 76 14688 76 16265 76 12553 76 12839 74 17400 74 17068 74 17419 74 13187 74 18260 74 13280 74 14210 74 16210 74 16367 74 14625 72 18077 72 13455 72 13448 72 13782 72 12356 72 12971 72 15332 70 17744 70 14748 70 14146 70 13385 70 15038 70 14696 70 16303 70 17964 70 12947 70 17213 70 13221 70 15513 70 18179 70 15301 68 12868 68 14396 68 14672 68 15291 68 16771 68 16907 68 12779 66 16898 66 17799 66 12621 66 13758 66 12708 66 16477 66 17812 66 16549 66 14051 66 18176 66 13126 64 13178 64 14256 64 14032 64 12395 64 12712 64 13426 64 15996 64 14896 64 17220 64 15696 62 13656 62 13001 62 12957 62 17913 62 14189 62 14258 62 15620 62 13520 62 14779 6

Or we can see examine those customers that operate in the most countries.

In [0]:
%sql
SELECT CustomerID, COUNT(Country) AS num_countries
FROM customer_counts
GROUP BY CustomerID
SORT BY num_countries DESC

CustomerID num_countries null 8 12370 2 14496 1 13090 1 14135 1 15260 1 17802 1 16995 1 13370 1 13269 1 13652 1 16858 1 12808 1 17218 1 13983 1 18259 1 14419 1 17236 1 15854 1 17303 1 14401 1 17525 1 13629 1 15046 1 16551 1 12651 1 12921 1 13777 1 15061 1 14031 1 16861 1 16782 1 14952 1 18269 1 15329 1 13000 1 18065 1 16153 1 17128 1 13694 1 12915 1 16926 1 16725 1 15332 1 13506 1 15973 1 15032 1 14437 1 16019 1 13148 1 14107 1 16927 1 16726 1 17551 1 12877 1 13029 1 15164 1 15862 1 13319 1 15024 1 15680 1 16083 1 16282 1 17389 1 13233 1 17800 1 17581 1 14625 1 12967 1 16814 1 16306 1 16365 1 17406 1 12766 1 15271 1 17671 1 13969 1 14639 1 14057 1 16503 1 13238 1 15727 1 16705 1 14045 1 15350 1 17855 1 17925 1 17961 1 12839 1 17975 1 16579 1 15044 1 17041 1 15062 1 15246 1 17850 1 16552 1 16781 1 17685 1 12725 1 16402 1 17442 1 15664 1 14414 1 16684 1 17420 1 17572 1 16770 1 13145 1 13013 1 16013 1 17198 1 18113 1 15021 1 15018 1 18061 1 12963 1 14159 1 17969 1 18141 1 13458 1 14426 1 15759 1 17848 1 17752 1 12763 1 15696 1 13502 1 12523 1 15213 1 18233 1 12451 1 12652 1 14587 1 14978 1 15447 1 16206 1 12490 1 13048 1 13527 1 14784 1 14657 1 15262 1 12433 1 16042 1 17211 1 14161 1 14344 1 16293 1 13599 1 17706 1 16385 1 13787 1 18044 1 17230 1 17744 1 16271 1 15464 1 13656 1 13747 1 16752 1 16916 1 15301 1 13769 1 16525 1 13523 1 16717 1 17430 1 18230 1 17880 1 17231 1 15101 1 12705 1 16270 1 13397 1 16424 1 12868 1 12779 1 13034 1 12681 1 17646 1 18004 1 17375 1 17449 1 12865 1 12666 1 16037 1 17888 1 17400 1 17849 1 16029 1 17920 1 13418 1 15955 1 14491 1 12686 1 17243 1 14708 1 17068 1 13126 1 17722 1 16553 1 14440 1 17978 1 13136 1 13324 1 12434 1 14373 1 14334 1 12944 1 15208 1 13082 1 14312 1 15100 1 13705 1 12855 1 12471 1 17567 1 16168 1 15426 1 15555 1 12474 1 16719 1 17869 1 14907 1 15002 1 15750 1 15805 1 15547 1 13890 1 14643 1 12515 1 15719 1 17095 1 12539 1 13117 1 17884 1 13089 1 16931 1 15880 1 15602 1 14176 1 17419 1 15628 1 13327 1 17696 1 14062 1 14878 1 18223 1 16945 1 13750 1 14761 1 12791 1 14142 1 15658 1 15732 1 16033 1 15894 1 14748 1 14085 1 13880 1 14796 1 15529 1 13982 1 16350 1 13178 1 14150 1 13266 1 16316 1 17580 1 14146 1 14553 1 14653 1 12748 1 17976 1 13958 1 14264 1 12738 1 15373 1 13174 1 16550 1 15079 1 14256 1 14543 1 17372 1 17633 1 16743 1 12348 1 16885 1 15527 1 14828 1 16326 1 17841 1 15502 1 13531 1 14684 1 13999 1 13284 1 18190 1 16948 1 15171 1 13993 1 16395 1 13385 1 16401 1 13929 1 14242 1 15518 1 12527 1 12736 1 15370 1 18069 1 12643 1 12626 1 14669 1 16500 1 13001 1 12402 1 14514 1 17862 1 13748 1 15738 1 14865 1 13807 1 12683 1 15856 1 14544 1 15159 1 17739 1 13496 1 17044 1 14841 1 16560 1 14533 1 15841 1 13187 1 15059 1 15093 1 15867 1 17324 1 16700 1 13211 1 16746 1 15204 1 12480 1 18074 1 13255 1 15012 1 14766 1 17381 1 14449 1 14236 1 16898 1 13495 1 15965 1 15660 1 14082 1 16656 1 17799 1 17158 1 12417 1 15465 1 15304 1 15298 1 17340 1 14243 1 12621 1 17339 1 16727 1 12875 1 17524 1 12476 1 16570 1 14524 1 15822 1 18260 1 12957 1 13280 1 16850 1 14047 1 17017 1 14388 1 16140 1 14396 1 13081 1 17858 1 16722 1 17616 1 18077 1 16011 1 17700 1 17146 1 14220 1 17913 1 12735 1 15611 1 17114 1 12583 1 17838 1 16905 1 14589 1 15769 1 15038 1 13094 1 16221 1 14448 1 17033 1 12577 1 14560 1 13680 1 15266 1 17897 1 14210 1 16327 1 15039 1 13481 1 18119 1 13786 1 16893 1 13140 1 16065 1 17179 1 16607 1 14911 1 17460 1 17732 1 14390 1 17228 1 12567 1 14409 1 17519 1 13756 1 16713 1 16016 1 14577 1 14646 1 18062 1 12503 1 18198 1 17457 1 15160 1 15808 1 14573 1 16686 1 15034 1 14487 1 17809 1 14040 1 17708 1 14572 1 13784 1 14320 1 12393 1 12601 1 12724 1 17609 1 13883 1 15749 1 12388 1 13064 1 15281 1 15189 1 15296 1 16791 1 12672 1 14916 1 12489 1 18085 1 17377 1 13108 1 17547 1 17320 1 18043 1 13077 1 15358 1 14205 1 17338 1 14913 1 16458 1 13758 1 14849 1 17690 1 16422 1 12586 1 14189 1 17191 1 12755 1 17411 1 14702 1 17790 1 17912 1 14961 1 16816 1 17204 1 17368 1 14258 1 17555 1 

And then look at how many orders a customer made in each of these countries.

In [0]:
%sql
SELECT Country, total_orders
FROM customer_counts
WHERE CustomerID = 20059

Country total_orders

### Using VACUUM to clean up small files

After we run OPTIMIZE, we have a number of uncompacted files that are no longer necessary. Running VACUUM will remove these files for us.

Let's go ahead and VACUUM our `customer_data_delta` table, which points at the files in our `DeltaPath` variable.

Count number of files before `VACUUM` for `Country=Sweden`.

In [0]:
# TODO
preNumFiles = len(dbutils.fs.ls(DeltaPath))
display(preNumFiles)

25

If you try to perform an immediate `VACUUM` (using `RETAIN 0 HOURS` to clean up recently optimized files), you will get an error.

In [0]:
%sql
-- TODO
VACUUM customer_data_delta RETAIN 0 HOURS;

---------------------------------------------------------------------------
IllegalArgumentException                  Traceback (most recent call last)
<command-1914720889771087> in <cell line: 1>()
      5     display(df)
      6     return df
----> 7   _sqldf = ____databricks_percent_sql()
      8 finally:
      9   del ____databricks_percent_sql

<command-1914720889771087> in ____databricks_percent_sql()
      2   def ____databricks_percent_sql():
      3     import base64
----> 4     df = spark.sql(base64.standard_b64decode("LS0gVE9ETwpWQUNVVU0gY3VzdG9tZXJfZGF0YV9kZWx0YSBSRVRBSU4gMCBIT1VSUw==").decode())
      5     display(df)
      6     return df

/databricks/spark/python/pyspark/instrumentation_utils.py in wrapper(*args, **kwargs)
     46             start = time.perf_counter()
     47             try:
---> 48                 res = func(*args, **kwargs)
     49                 logger.log_success(
     50                     module_name, class_name, function_name, time.perf_coun

This is a helfpul error. Remember that `VACUUM` is intended for occasional garbage collection. Here we'll just demonstrating that we _can_ use it to clean up files, so we'll set our configuration to allow this operation.

In [0]:
spark.conf.set("spark.databricks.delta.retentionDurationCheck.enabled", False)

Now we won't get an error when we run `VACUUM`.

In [0]:
%sql
VACUUM customer_data_delta RETAIN 0 HOURS;

path dbfs:/user/saifahmed.k@outlook.com/delta/customer-data

Count how many files there are for `Country=Sweden`.

In [0]:
# TODO
postNumFiles = len(dbutils.fs.ls(DeltaPath))
display(postNumFiles)

25

Comparing our `preNumFiles` to `postNumFiles`, we can see that this number has reduced.

This notebook is focused on optimizing and managing data in a **Delta Lake**, which is an open-source storage layer that helps with efficient data processing and version control in data lakes.

Here’s a breakdown of the key concepts covered:

### 1. **Delta Lake**
   - **Delta Lake** is a tool that allows you to store large amounts of data in a way that supports version control and transactional reliability (ACID compliance). It enables efficient **reading**, **writing**, and **querying** of data stored in data lakes.

### 2. **Time Travel**
   - **Time Travel** in Delta Lake allows you to look at different **versions** of your data at specific points in time. This means you can query data as it existed at any version by using the **`VERSION AS OF`** feature.
   - Example: You query an older version of a table to see how the data looked in the past or compare changes over time.

### 3. **Optimization (OPTIMIZE & ZORDER)**
   - **Optimize**: In a Delta Lake, data is stored in multiple small files. Over time, as data is updated, the number of these small files can grow, which can slow down query performance. 
     - **Optimization** compacts these small files into larger ones, making the system faster and more efficient to query.
   - **ZORDER**: This is a method of sorting data to further speed up specific types of queries.
     - For example, if you often query data based on a specific column (e.g., **CustomerID**), you can reorder the data using **ZORDER** to make the searches more efficient by reducing the number of files that need to be checked.

### 4. **VACUUM**
   - **Vacuum** is a process that cleans up **old and unnecessary files** in Delta Lake.
   - After optimizing or making changes, there can be leftover data files that are no longer needed. **VACUUM** removes these old files, freeing up storage space and making the data system more efficient.
   - Important: Delta Lake retains old versions of data files for a set time (default is 7 days) to support time travel. If you try to vacuum immediately, it may throw an error because recent files may still be needed. You can override this by adjusting the retention setting.

### 5. **Partitioning**
   - **Partitioning** divides data into subsets based on a specific column, such as **Country** or **CustomerID**. This allows queries to focus on a specific subset of data, improving performance.
   - Example: Partitioning data by country can help when you want to query all data from **Sweden** without needing to scan data from other countries.

### 6. **Aggregations**
   - **Aggregations** (like SUM or COUNT) are calculations across your data. For instance, you can calculate the total number of orders from a specific country or customer.
   - This notebook demonstrates using SQL to calculate such aggregates and compare data between different time versions (before and after changes).

### 7. **SQL Queries**
   - You interact with the Delta Lake data using **SQL queries** to retrieve or manipulate data.
   - Examples in the notebook include querying the total number of orders, retrieving historical data (using `VERSION AS OF`), and running optimization commands.

### 8. **Cluster Configuration**
   - Spark is used to process the data in parallel using multiple nodes (computers). In the notebook, the number of partitions is reduced to better match the available resources of the cluster, improving efficiency.

### Summary:
This notebook helps you manage and optimize large datasets stored in Delta Lake. It allows you to:
- **View historical data versions** (Time Travel)
- **Optimize performance** by reducing the number of small files and sorting data efficiently (OPTIMIZE and ZORDER)
- **Clean up unnecessary files** after updates (VACUUM)
- **Query data** efficiently using SQL, leveraging features like partitioning to speed up queries.

Let's break down **OPTIMIZE** and **VACUUM** using simple examples with tables.

### 1. **OPTIMIZE**
**OPTIMIZE** is used to compact many small files in a Delta table into fewer, larger files. This makes queries run faster because it reduces the number of files Spark has to read.

#### Example Scenario:
Imagine you have a Delta table called `customer_orders` that stores millions of customer orders. Each update or write operation might create many small files. Over time, as more data is added, the number of small files grows, and queries start to slow down because the system has to read a lot of files to retrieve the data.

**Without Optimization**:
```plaintext
Table: customer_orders

Files:
order_part_001.parquet
order_part_002.parquet
order_part_003.parquet
...
order_part_5000.parquet
```
You might have thousands of small files, and every time you query the table (e.g., to get the total orders for a customer), Spark has to scan all those files.

Now, let’s **optimize** this table to improve query performance.

**Optimization Command**:
```sql
OPTIMIZE customer_orders
```

What happens:
- The system will **compact** these files. It will merge smaller files into larger ones, reducing the overall number of files.
  
**After Optimization**:
```plaintext
Table: customer_orders

Files:
order_part_optimized_001.parquet
order_part_optimized_002.parquet
```
Now, instead of thousands of small files, we have just a few large ones, making the queries faster.

#### ZORDER for Even Faster Queries:
You can further optimize the table based on frequently queried columns. For instance, if you often query `customer_orders` by **CustomerID**, you can use **ZORDER** to reorder the data.

**Command with ZORDER**:
```sql
OPTIMIZE customer_orders ZORDER BY (CustomerID)
```
This ensures that rows with the same `CustomerID` are stored together in fewer files, which improves the speed of queries related to `CustomerID`.

---

### 2. **VACUUM**
**VACUUM** cleans up old, unnecessary files that are no longer in use after updates or optimizations. Delta Lake keeps old files to support features like time travel, but over time, these files take up space and are not needed anymore.

#### Example Scenario:
After you update or optimize the `customer_orders` table, there might be old versions of files lying around, which are no longer part of the current table. These old files consume storage space.

Before running **VACUUM**, your table might look like this:

**Files After Updates (Before VACUUM)**:
```plaintext
Files:
order_part_001.parquet    <-- Old file (no longer needed)
order_part_002.parquet    <-- Old file (no longer needed)
order_part_optimized_001.parquet  <-- New file (current data)
order_part_optimized_002.parquet  <-- New file (current data)
```

The old files `order_part_001.parquet` and `order_part_002.parquet` are no longer used in the current table after optimization, but they still exist for time travel or rollback purposes.

To clean up these files, you can run the **VACUUM** command.

**Vacuum Command**:
```sql
VACUUM customer_orders RETAIN 168 HOURS
```

This command tells Delta Lake to delete files that are older than **168 hours** (7 days). You can adjust the number of hours if you want to keep the old files for a shorter or longer period. **After VACUUM**, only the necessary, optimized files remain:

**Files After VACUUM**:
```plaintext
Files:
order_part_optimized_001.parquet  <-- New file (current data)
order_part_optimized_002.parquet  <-- New file (current data)
```

In this case, the old files are removed, freeing up storage space.

---

### Summary with a Simple Analogy:
- **OPTIMIZE** is like cleaning up your workspace by putting all scattered papers (small files) into a few neatly organized folders (larger, compacted files).
- **VACUUM** is like throwing away old drafts or documents you no longer need after reorganizing your workspace.

Together, these operations keep your Delta Lake tables efficient and clean, speeding up query performance and saving storage space.